### PR TITLE
Add linking events possibility in stream

### PR DESCRIPTION
**Pull request by aroeczek-u**
https://yousty.atlassian.net/browse/YWEB-8391

If we agree for the solution I'll update documentation

### CODE REVIEW

File: lib/event_store_client/event.rb [hunk 1]
```diff
@@ -8,9 +8,10 @@ module EventStoreClient
   class Event < Dry::Struct
     attr_reader :id
 
+    attribute :type, Types::Strict::String
+    attribute :title, Types::Strict::String.optional.default(nil)
```
Comment (swilgosz) [inline lib/event_store_client/event.rb]: Does it work? `Types::Strict::String` should return error for nil. But maybe for optional it's not a case. Just please confirm if I'm wrong.

File: lib/event_store_client/deserialized_event.rb [hunk 2]
```diff
@@ -6,9 +6,11 @@ module EventStoreClient
   class DeserializedEvent
     InvalidDataError = Class.new(StandardError)
 
+    attr_reader :id
+    attr_reader :type
+    attr_reader :title
     attr_reader :data
     attr_reader :metadata
-    attr_reader :type
```
Comment (swilgosz) [inline lib/event_store_client/deserialized_event.rb]: why did you move the type above? I'd understand if you'd order attrs alphabetically, but the data and metadata are last items here, so I'm not sure what was the reason.

File: lib/event_store_client/store_adapter/in_memory.rb [hunk 3]
```diff
@@ -19,18 +19,20 @@ def append_to_stream(stream_name, events, expected_version: nil) # rubocop:disab
...
+
+        read_stream_backward(stream_name, start: start, count: count)
       end
 
       def delete_stream(stream_name, hard_delete: false) # rubocop:disable Lint/UnusedMethodArgument
         event_store.delete(stream_name)
       end
 
+      def link_to(stream_name, events)
+        raise NotImplementedError
```
Comment (swilgosz) [inline lib/event_store_client/store_adapter/in_memory.rb]: Don't you need that for testing?

It's easy - just add the events under new stream key - with the same ID and everything. Here we don't care about duplications, do we?

File: spec/event_store_client/client_spec.rb [hunk 4]
```diff
@@ -47,6 +47,47 @@ module EventStoreClient
       end
     end
 
+    describe 'link_to' do
```
Comment (swilgosz) [inline spec/event_store_client/client_spec.rb]: Please follow the naming convention when naming describe blocks for methods: 

describe "#METHOD_NAME" - for instance mehtods
describe ".METHOD_NAME" - for class methods

look above for examples

File: spec/event_store_client/client_spec.rb [hunk 5]
```diff
@@ -47,6 +47,47 @@ module EventStoreClient
...
+          expect{ subject.call }.to raise_error(ArgumentError)
+        end
+      end
+
+      context 'when missing stream' do
+        let(:stream_name) { nil }
+        it_behaves_like 'argument error'
+      end
+
+      context 'when missing events' do
```
Comment (swilgosz) [inline spec/event_store_client/client_spec.rb]: Missing test for let(:events) { event } - it will crash with unexpected error (undefined_method :empty?)

File: spec/event_store_client/store_adapter/api/client_spec.rb [hunk 6]
```diff
@@ -94,9 +96,56 @@ module EventStoreClient::StoreAdapter::Api
...
+
+    describe '#link_to' do
+      let(:event_1) do
+        EventStoreClient::Event.new(
+          id: SecureRandom.uuid,
+          title: '3@Stream',
+          type: 'SomethingHappened'
+        )
+      end
+      let(:event_2) do
```
Comment (swilgosz) [inline spec/event_store_client/store_adapter/api/client_spec.rb]: Missing empty line above

File: spec/event_store_client/store_adapter/api/client_spec.rb [hunk 7]
```diff
@@ -94,9 +96,56 @@ module EventStoreClient::StoreAdapter::Api
...
+          with(body: body.to_json)
+      end
+
+      it 'sends a correct request' do
+        api_client.link_to(stream_name, events)
         expect(stub).to have_been_requested
       end
     end
+
+    private
```
Comment (swilgosz) [inline spec/event_store_client/store_adapter/api/client_spec.rb]: Private inside of `Rspec.describe?` does it have effect?

File: lib/event_store_client/connection.rb [hunk 8]
```diff
@@ -51,6 +52,12 @@ def consume_feed(stream, subscription)
       events
     end
 
+    def link_to(stream, events)
+      client.link_to(stream, events)
+
```
[the remark this answers is not in the document]
Comment (aroeczek-u) [reply lib/event_store_client/connection.rb]: To be clearly separated returned value from logic

File: lib/event_store_client/event.rb (hunk 1 above)
Comment (aroeczek-u) [reply lib/event_store_client/event.rb]:
```
EventStoreClient::Event.new(type: 'MyType')     
 => #<EventStoreClient::Event type="MyType" title=nil data="{}" metadata="{\"created_at\":\"2019-12-27 15:37:34 +0100\"}"> 
```

File: lib/event_store_client/deserialized_event.rb (hunk 2 above)
Comment (aroeczek-u) [reply lib/event_store_client/deserialized_event.rb]: For me make sense to have id than type than attributes than data than metadata. But also you are also right that makes sense to have ordered alphabetically. So do you think we should keep alphabetical order?

File: lib/event_store_client/store_adapter/in_memory.rb (hunk 3 above)
Comment (aroeczek-u) [reply lib/event_store_client/store_adapter/in_memory.rb]: I didn't even think that is so super easy to implement. So I just did it.

File: spec/event_store_client/store_adapter/api/client_spec.rb (hunk 7 above)
Comment (aroeczek-u) [reply spec/event_store_client/store_adapter/api/client_spec.rb]: Honestly have no clue but I was always put custom methods in spec under private.

File: lib/event_store_client/client.rb [hunk 9]
```diff
@@ -61,6 +61,12 @@ def stop_polling
       nil
     end
 
+    def link_to(stream:, events:)
+      raise ArgumentError if !stream || stream == '' || events.empty?
```
[the remark this answers is not in the document]
Comment (aroeczek-u) [reply lib/event_store_client/client.rb]: Resolved we can pass array of events or single event.

File: lib/event_store_client/connection.rb [hunk 10]
```diff
@@ -51,6 +52,12 @@ def consume_feed(stream, subscription)
...
+      events
```
[the remark this answers is not in the document]
Comment (aroeczek-u) [reply lib/event_store_client/connection.rb]: Fixed

File: spec/event_store_client/client_spec.rb (hunk 5 above)
Comment (aroeczek-u) [reply spec/event_store_client/client_spec.rb]: Added

File: spec/event_store_client/store_adapter/api/client_spec.rb (hunk 6 above)
Comment (aroeczek-u) [reply spec/event_store_client/store_adapter/api/client_spec.rb]: Fixed

**Comment by aroeczek-u** [comment]
> Most of my comments are trivial lint/syntax improvement suggestions, but there were few minor thing that needs to be fixed.

Thanks for Review I've just introduced improvements and responded to your comments.
